### ISSUE TITLE
Lwt fit

### ISSUE THREAD

**Issue by key23** [ocsigen/lwt#333]
I'm having a tough time understanding Lwt fit and usage. 

Lwt has a small core api. Used for spawning and using threads. I'm not quite sure why there is also sub modules for IO, Unix, and Sys. Especially when those submodules don't encompass all of the standard libraries functionality as in there are missing operators. What I'm trying to say is, can't we just use the Lwt core to wrap standard library IO and pass it to bind?

**Comment by key23** [comment]
For example, instead of using Lwt_io would the code below be non blocking if wrapped in bind?
```ocaml
let stuff = Lwt.return 42
let _ = Lwt.bind stuff (fun int  -> print_int int; Lwt.return int)
```

**Comment by c-cube** [comment]
No, this is blocking, because the standard `print_int` is blocking. If it were to block for a while (e.g. because the output is redirected to some network socket) then the whole program would block.

**Comment by key23** [comment]
@c-cube how would I know what is blocking and what is not blocking?

**Comment by rgrinberg** [comment]
General rules of thumb:

* anything that returns `'a Lwt.t` isn't blocking.
* This is why all those `Lwt_*` modules you pointed out exist. They provide non blocking versions of the stdlib.

**Comment by aantron** [comment]
First, `'a Lwt.t`s aren't *threads*, they are *promises*. I know the manual refers to them as threads throughout. It's an outstanding [task][manual] to fix that. Sorry for the confusion.

To make the above answers a bit more concrete:

1. This is blocking, because `Unix.openfile` internally directly calls the `open(2)` system call, which can block:

    ```ocaml
    Unix.openfile "./foo.txt" [Unix.OWRONLY] 0
    ```

    and the result is of type `Unix.file_descr`.

2. This is non-blocking, because `Lwt_unix.openfile` calls the same `open(2)` system call on a hidden worker system thread, which Lwt manages for you:

    ```ocaml
    Lwt_unix.openfile "./foo.txt" [Unix.WRONLY] 0
    ```

    This function returns immediately (often before `open(2)` returns). Its result is of type `Lwt_unix.file_descr Lwt.t`, which is a promise for a `Lwt_unix.file_descr`. When the `open(2)` does finally complete inside that hidden worker thread, Lwt resolves the promise for you by providing the actual `Lwt_unix.file_descr` value, which you can take out with `bind`:

    ```ocaml
    let fd_promise = Lwt_unix.openfile "./foo.txt" [Unix.WRONLY] 0 in
    Lwt.bind fd_promise (fun fd ->
        (* when this eventually gets called, we have an actual fd *))
    ```

3. You *can* mix Lwt and blocking calls, however the result will be blocking I/O. This code still blocks as in (1):

    ```ocaml
    Lwt.return (Unix.openfile "./foo.txt" [Unix.WRONLY] 0)
    ```

    Even though the result type is `Unix.file_descr Lwt.t`, the code is actually equivalent to

    ```ocaml
    let fd = Unix.openfile "./foo.txt" [Unix.WRONLY] 0 in
    Lwt.return fd
    ```

    In other words, first it runs a blocking operation to get `fd`, and then wraps that in a trivial, already-resolved promise using `Lwt.return`. Even if you `bind` on it later, you've still already done that blocking call:

    ```ocaml
    Lwt.bind
      (Lwt.return (Unix.openfile "./foo.txt" [Unix.WRONLY] 0))
      (fun fd -> (* ... *))
    ```

    is equivalent to

    ```ocaml
    let fd = Unix.openfile "./foo.txt" [Unix.WRONLY] 0 in    (*  :'(  *)
    let fd_promise = Lwt.return fd in
    Lwt.bind fd_promise (fun fd -> (* ... *))
    ```

[manual]: https://github.com/ocsigen/lwt/issues/304

**Comment by aantron** [comment]
As for why Lwt doesn't have its own version of everything in `Unix`, etc., usually a function is missing for one of these reasons:

- The function can't block anyway, e.g. `Unix.gettimeofday ()`. You should just use the `Unix` one then.
- The function can be easily implemented in terms of other functions in Lwt, though I don't have an example handy.
- The function is too platform-specific.
- It's an oversight :)

Lwt is extensible by other libraries. For example, [ocaml-unix-unistd](https://github.com/dsheets/ocaml-unix-unistd) provides, among other things, Lwt-friendly non-blocking bindings to `pread` and `pwrite`.

**Comment by key23** [comment]
Awesome thanks!

**Comment by rgrinberg** [comment]
Anton, you should start a FAQ to icnlude these great answers.

**Comment by aantron** [comment]
@rgrinberg Something on this topic is actually in the new manual draft, but at this rate, I don't know when that will be done. It is still highly incomplete, as I got diverted into writing tests and reading the code deeply to figure out what Lwt really does :) So maybe a FAQ is a good idea.
